File: packages/protocol/src/aead.ts
/**
 * AES-256-GCM frame codec (plan.md §6.3). Each frame is a fixed 16-byte header followed
 * by the GCM output (ciphertext with appended 16-byte tag). The encoded header is the
 * GCM additional authenticated data, so any header tampering fails decryption.
 *
 * The 12-byte nonce is the direction's 4-byte salt followed by a big-endian u64 frame
 * counter. Counters are per-direction and supplied by the caller (the session tracks
 * them), giving nonce continuity without ever reusing a (key, nonce) pair. Uses native
 * WebCrypto; mirrored by Go `crypto/cipher` in `packages/wire`.
 */

import { AEAD_NONCE_BYTES, AEAD_TAG_BYTES, FRAME_HEADER_BYTES } from './constants.js';
import { decodeFrameHeader, encodeFrameHeader } from './frame.js';
import type { DirectionalKey } from './keyschedule.js';
import type { FrameHeader } from './transfer.js';
import { aesGcmOpen, aesGcmSeal } from './webcrypto.js';

/** Header fields the caller supplies; `len` is set from the plaintext by `seal`. */
export type FrameHeaderInput = Omit<FrameHeader, 'len'>;

const U16_MAX = 0xffff;

function nonce(salt: Uint8Array, counter: number): Uint8Array {
  if (!Number.isInteger(counter) || counter < 0) {
    throw new RangeError(`frame counter ${counter} is not a non-negative integer`);
  }
  const out = new Uint8Array(AEAD_NONCE_BYTES);
  out.set(salt, 0);
  new DataView(out.buffer).setBigUint64(salt.length, BigInt(counter), false);
  return out;
}

/**
 * Encrypt `plaintext` into a frame: `header(16) || ciphertext || tag(16)`. The header's
 * `len` field is set to the plaintext length and authenticated as GCM AAD.
 */
export async function seal(
  dir: DirectionalKey,
  counter: number,
  header: FrameHeaderInput,
  plaintext: Uint8Array,
): Promise<Uint8Array> {
  if (plaintext.length > U16_MAX) {
    throw new RangeError(`frame payload ${plaintext.length} exceeds u16 max ${U16_MAX}`);
  }
  const aad = encodeFrameHeader({ ...header, len: plaintext.length });
  const sealed = await aesGcmSeal(dir.key, nonce(dir.salt, counter), aad, plaintext);
  const out = new Uint8Array(aad.length + sealed.length);
  out.set(aad, 0);
  out.set(sealed, aad.length);
  return out;
}

/** A decrypted frame: its parsed header and recovered plaintext. */
export interface OpenedFrame {
  readonly header: FrameHeader;
  readonly plaintext: Uint8Array;
}

/**
 * Decrypt a frame produced by `seal`. Verifies the GCM tag over the header AAD and the
 * expected nonce; throws if authentication fails, the frame is truncated, or the header
 * `len` disagrees with the ciphertext length.
 */
export async function open(
  dir: DirectionalKey,
  counter: number,
  frame: Uint8Array,
): Promise<OpenedFrame> {
  if (frame.length < FRAME_HEADER_BYTES + AEAD_TAG_BYTES) {
    throw new Error(`frame too short: ${frame.length} bytes`);
  }
  const aad = frame.slice(0, FRAME_HEADER_BYTES);
  const header = decodeFrameHeader(aad);
  const body = frame.slice(FRAME_HEADER_BYTES);
  if (body.length !== header.len + AEAD_TAG_BYTES) {
    throw new Error(`frame len field ${header.len} disagrees with body ${body.length}`);
  }
  const plaintext = await aesGcmOpen(dir.key, nonce(dir.salt, counter), aad, body);
  return { header, plaintext };
}

File: packages/wire/aead.go
package wire

import (
	"crypto/aes"
	"crypto/cipher"
	"encoding/binary"
	"fmt"
)

const u16Max = 0xffff

// FrameHeaderInput is the header fields the caller supplies; Len is set from the
// plaintext by Seal.
type FrameHeaderInput struct {
	Version  uint8
	Type     uint8
	Flags    uint8
	FileIdx  uint16
	BlockIdx uint32
	FrameOff uint32
}

// nonce builds the 12-byte GCM nonce: the direction's 4-byte salt followed by a
// big-endian u64 frame counter.
func nonce(salt []byte, counter uint64) []byte {
	out := make([]byte, aeadNonceBytes)
	copy(out, salt)
	binary.BigEndian.PutUint64(out[aeadSaltBytes:], counter)
	return out
}

func gcmFor(key []byte) (cipher.AEAD, error) {
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, err
	}
	return cipher.NewGCM(block)
}

// Seal encrypts plaintext into a frame: header(16) || ciphertext || tag(16). The header's
// Len field is set to the plaintext length and authenticated as GCM additional data.
func Seal(dir DirectionalKey, counter uint64, h FrameHeaderInput, plaintext []byte) ([]byte, error) {
	if len(plaintext) > u16Max {
		return nil, fmt.Errorf("frame payload %d exceeds u16 max %d", len(plaintext), u16Max)
	}
	aad := encodeFrameHeader(FrameHeader{
		Version:  h.Version,
		Type:     h.Type,
		Flags:    h.Flags,
		FileIdx:  h.FileIdx,
		BlockIdx: h.BlockIdx,
		FrameOff: h.FrameOff,
		Len:      uint16(len(plaintext)),
	})
	gcm, err := gcmFor(dir.Key)
	if err != nil {
		return nil, err
	}
	// Seal appends ciphertext||tag onto aad, giving header || ciphertext || tag.
	return gcm.Seal(aad, nonce(dir.Salt, counter), plaintext, aad), nil
}

// OpenedFrame is a decrypted frame: its parsed header and recovered plaintext.
type OpenedFrame struct {
	Header    FrameHeader
	Plaintext []byte
}

// Open decrypts a frame produced by Seal. It verifies the GCM tag over the header AAD and
// the expected nonce, and fails if authentication fails, the frame is truncated, or the
// header Len disagrees with the ciphertext length.
func Open(dir DirectionalKey, counter uint64, frame []byte) (*OpenedFrame, error) {
	if len(frame) < frameHeaderBytes+aeadTagBytes {
		return nil, fmt.Errorf("frame too short: %d bytes", len(frame))
	}
	aad := frame[:frameHeaderBytes]
	header, err := decodeFrameHeader(aad)
	if err != nil {
		return nil, err
	}
	body := frame[frameHeaderBytes:]
	if len(body) != int(header.Len)+aeadTagBytes {
		return nil, fmt.Errorf("frame len field %d disagrees with body %d", header.Len, len(body))
	}
	gcm, err := gcmFor(dir.Key)
	if err != nil {
		return nil, err
	}
	plaintext, err := gcm.Open(nil, nonce(dir.Salt, counter), body, aad)
	if err != nil {
		return nil, err
	}
	return &OpenedFrame{Header: header, Plaintext: plaintext}, nil
}
